dotnet/sdk-phase-1.md

```diff
@@ -0,0 +1,1022 @@
+    activity
+  * Not required for static activities because they can just be referenced on the class
+* Any overridden instance activities from base classes or same-named static activities from base classes must have the
+  exact same signature and attribute. This ensures people don't make mistakes accidentally overriding and removing the
+  activity attribute. This also avoids a diamond problem or duplicate problem for statics.
+
+### Activity Contextual Operations
+
+The activity context looks like this:
+
+```csharp
+namespace Temporal.Activity;
+
+public class ActivityContext
+{
+    public static ActivityContext Current { get; }
+
+    public static bool HasCurrent { get; }
+
+    internal ActivityContext(
+        ActivityInfo info,
+        CancellationToken cancellationToken,
+        Action<object[]> heartbeater) { /*...*/ }
+
+    public ActivityInfo Info { get; }
+
+    public CancellationToken CancellationToken { get; }
+
+    public CancellationToken WorkerShutdownToken { get; }
+
+    internal Action<object[]> Heartbeater { get; }
+
+    public void Heartbeat(params object[] details) { /*...*/ }
+}
+```
+
+Notes:
+
+* The current activity context is on thread local or async local depending on whether activity is async or not
+* All calls except `HasCurrent` fail if not in activity context
+* `ActivityInfo` is an immutable record containing info
+* `CancellationToken` and `WorkerShutdownToken` are usable by sync and async
+  * `WorkerShutdownToken` is set on worker shutdown before cancellation token is set. If there is no graceful shutdown
+    period, one is set immediately after the other
+* Note the lack of logger here. See "Logging" later, but in general .NET doesn't have a logger in the standard library.
+  We may make some kind of `IReadOnlyDictionary<string, string> ActivityInfo.LogInfo { get; }` that can be used with
+  `ILogger.BeginScope` or similar.
+
+### Async Completion
+
+Simply `throw new Temporal.Activity.CompleteAsyncException` from the activity.
+
+## Client
+
+Inside the `Temporal.Client` namespace, we have the following types:
+
+* `TemporalClient` - namespace-specific, stateless high-level client to Temporal
+  * See later for most of its API design
+  * Accepts `ITemporalConnection`
+  * Implements `ITemporalClient` for DI use (which itself implements `Temporal.Worker.IWorkerClient` for worker use)
+* `TemporalConnection` - namespace-agnostic, stateful low-level client to Temporal
+  * Has `WorkflowService`, `OperatorService`, etc for making raw gRPC calls
+  * Supports lazy connectivity
+  * Implements `ITemporalConnection` for DI use
+* `WorkflowHandle` - high-level workflow-specific calls
+* Many other supporting types
+
+Example of starting the workflow defined earlier in document, signalling it, then waiting on its results:
+
+```csharp
+using Temporal.Client;
+using MyNamespace;
+
+// Connect
+var client = await TemporalClient.ConnectAsync(new("localhost:7233"));
```

Comment:
Not by default. I don't believe it is our standard. We encourage people to connect eagerly to find connection issues before their first call (or before their worker just hangs). However, `Lazy` is an option here, but I call it `ConnectAsync` (like I call it `connect` in Python) because it connects by default. If I had my way we wouldn't even support lazy :-)

Reply:
Why accept a `Lazy` option instead of a static method with that name and avoid the runtime error altogether?

Reply:
We need a common entry point for lazy and non-lazy here as we did in Python IMO. This is even more important for .NET where people will use configuration separately from where they make the call (https://learn.microsoft.com/en-us/dotnet/core/extensions/configuration will work with our options by intention).

Also, lazy should be rare/discouraged IMO. And we have to carry it around as an option anyways. Having one entry point to client connection is clearer than overloads.

It's only a runtime error if they try to use the connection constructor in a non-lazy way. I think it's worth always having the constructor available for those wanting to manually construct a connection though (unless there's a good reason, constructors are often made visible to users in .NET).